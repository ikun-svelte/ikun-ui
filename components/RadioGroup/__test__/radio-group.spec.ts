import { tick } from 'svelte';
import { afterEach, expect, test, describe, beforeEach, vi } from 'vitest';
import KRadioGroup from '../src';
import KGroupBase from './fixture/group.base.svelte';
import KGroupChange from './fixture/group.change.svelte';
import KGroupTrigger from './fixture/group.trigger.svelte';
import KGroupBaseDisabled from './fixture/group.base.disabled.svelte';
import KGroupChangeDisabled from './fixture/group.change.disabled.svelte';
import KGroupTriggerDisabled from './fixture/group.trigger.disabled.svelte';
import KGroupModelDisabled from './fixture/group.model.disabled.svelte';
import KGroupSize from './fixture/group.size.svelte';

let host: HTMLElement;

const initHost = () => {
	host = document.createElement('div');
	host.setAttribute('id', 'host');
	document.body.appendChild(host);
};
beforeEach(() => {
	initHost();
	vi.useFakeTimers();
});
afterEach(() => {
	host.remove();
	vi.restoreAllMocks();
});

describe('Test: KRadioGroup', () => {
	test('props: cls', async () => {
		const instance = new KRadioGroup({
			target: host,
			props: {
				cls: 'k-radio-group--test'
			}
		});
		expect(instance).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('k-radio-group--test')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: size', async () => {
		const instance = new KGroupSize({
			target: host
		});
		await tick();
		expect(instance).toBeTruthy();
		const smEls = host.querySelector('#group-sm');
		expect(smEls.querySelectorAll('.k-radio--sm').length).toBe(3);
		const mdEls = host.querySelector('#group-md');
		expect(mdEls.querySelectorAll('.k-radio--md').length).toBe(5);
		const lgEls = host.querySelector('#group-lg');
		expect(lgEls.querySelectorAll('.k-radio--lg').length).toBe(2);
		expect(host.innerHTML).matchSnapshot();
	});

	test('When it is not disabled, the value of the radio can be initialized.', async () => {
		const instance = new KGroupBase({
			target: host
		});
		await tick();
		expect(instance).toBeTruthy();
		const inputEls = host.querySelectorAll('input');
		expect(inputEls[0].value === 'false').toBeTruthy();
		expect(inputEls[1].value === 'true').toBeTruthy();
		expect(inputEls[2].value === 'false').toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('When it is not disabled, the value changes and the status of the radio changes.', async () => {
		const instance = new KGroupChange({
			target: host
		});
		await tick();
		expect(instance).toBeTruthy();
		const inputEls = host.querySelectorAll('input');
		expect(inputEls[0].value === 'false').toBeTruthy();
		expect(inputEls[1].value === 'false').toBeTruthy();
		expect(inputEls[2].value === 'true').toBeTruthy();
		const btn = host.querySelector('#group_change_btn');
		btn.click(); // or btn.dispatchEvent(new window.Event('click', { bubbles: true }))
		await tick();
		expect(inputEls[0].value === 'true').toBeTruthy();
		expect(inputEls[1].value === 'false').toBeTruthy();
		expect(inputEls[2].value === 'false').toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('When it is not disabled, when a value of the radio changes, the group value is triggered.', async () => {
		const instance = new KGroupTrigger({
			target: host
		});
		await tick();
		expect(instance).toBeTruthy();
		const inputEls = host.querySelectorAll('input');
		expect(inputEls[0].value === 'false').toBeTruthy();
		expect(inputEls[1].value === 'false').toBeTruthy();
		expect(inputEls[2].value === 'true').toBeTruthy();
		const divEl = host.querySelector('#group_trigger_value');
		expect(divEl?.innerHTML).toBe('3');
		inputEls[0].click();
		await tick();
		expect(divEl?.innerHTML).toBe('1');
		expect(inputEls[0].value === 'true').toBeTruthy();
		expect(inputEls[1].value === 'false').toBeTruthy();
		expect(inputEls[2].value === 'false').toBeTruthy();
		inputEls[1].click();
		await tick();
		expect(divEl?.innerHTML).toBe('2');
		expect(host.innerHTML).matchSnapshot();
	});

	test('When it is disabled, the value of the radio can be initialized.', async () => {
		const instance = new KGroupBaseDisabled({
			target: host
		});
		await tick();
		expect(instance).toBeTruthy();
		const inputEls = host.querySelectorAll('input');
		expect(inputEls[0].value === 'true').toBeTruthy();
		expect(inputEls[1].value === 'false').toBeTruthy();
		expect(inputEls[2].value === 'false').toBeTruthy();
		expect(host.innerHTML.includes('k-radio--box__disabled')).toBeTruthy();
		expect(host.innerHTML.match(/k-radio--box__disabled/g)?.length === 3).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('When it is disabled, the value changes and the status of the radio not changes.', async () => {
		const instance = new KGroupChangeDisabled({
			target: host
		});
		await tick();
		expect(instance).toBeTruthy();
		const inputEls = host.querySelectorAll('input');
		expect(inputEls[0].value === 'false').toBeTruthy();
		expect(inputEls[1].value === 'false').toBeTruthy();
		expect(inputEls[2].value === 'true').toBeTruthy();
		const btn = host.querySelector('#group_change_disabled_btn');
		btn.click(); // or btn.dispatchEvent(new window.Event('click', { bubbles: true }))
		await tick();
		expect(inputEls[0].value === 'false').toBeTruthy();
		expect(inputEls[1].value === 'false').toBeTruthy();
		expect(inputEls[2].value === 'true').toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('When it is disabled, when a value of the radio changes, the group value is not triggered.', async () => {
		const instance = new KGroupTriggerDisabled({
			target: host
		});
		await tick();
		expect(instance).toBeTruthy();
		const inputEls = host.querySelectorAll('input');
		expect(inputEls[0].value === 'false').toBeTruthy();
		expect(inputEls[1].value === 'false').toBeTruthy();
		expect(inputEls[2].value === 'true').toBeTruthy();
		const divEl = host.querySelector('#group_trigger_value');
		expect(divEl?.innerHTML).toBe('3');
		inputEls[0].click();
		await tick();
		expect(divEl?.innerHTML).toBe('3');
		expect(inputEls[0].value === 'false').toBeTruthy();
		expect(inputEls[1].value === 'false').toBeTruthy();
		expect(inputEls[2].value === 'true').toBeTruthy();
		inputEls[1].click();
		await tick();
		expect(divEl?.innerHTML).toBe('3');
		inputEls[0].click();
		await tick();
		inputEls[1].click();
		await tick();
		inputEls[2].click();
		await tick();
		expect(divEl?.innerHTML).toBe('3');
		expect(host.innerHTML).matchSnapshot();
	});

	test('When the disabled value changes, the radio value can be synchronized.', async () => {
		const instance = new KGroupModelDisabled({
			target: host
		});
		await tick();
		expect(instance).toBeTruthy();
		const inputEls = host.querySelectorAll('input');
		expect(inputEls[0].value === 'false').toBeTruthy();
		expect(inputEls[1].value === 'false').toBeTruthy();
		expect(inputEls[2].value === 'true').toBeTruthy();
		const btn = host.querySelector('#group_change_value_btn');
		btn.click(); // or btn.dispatchEvent(new window.Event('click', { bubbles: true }))
		await tick();
		expect(inputEls[0].value === 'false').toBeTruthy();
		expect(inputEls[1].value === 'false').toBeTruthy();
		expect(inputEls[2].value === 'true').toBeTruthy();

		const btn2 = host.querySelector('#group_change_disabled_btn');
		btn2.click(); // or btn.dispatchEvent(new window.Event('click', { bubbles: true }))
		await tick();
		expect(inputEls[0].value === 'true').toBeTruthy();
		expect(inputEls[1].value === 'false').toBeTruthy();
		expect(inputEls[2].value === 'false').toBeTruthy();

		expect(host.innerHTML).matchSnapshot();
	});
});
