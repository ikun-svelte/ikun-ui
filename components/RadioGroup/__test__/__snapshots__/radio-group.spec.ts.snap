// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test: KRadioGroup > When it is disabled, the value changes and the status of the radio not changes. 1`] = `"<div class=\\"h-full\\" style=\\"width: 600px;\\"><button id=\\"group_change_disabled_btn\\"></button> <div class=\\"k-radio-group k-radio-group--base\\"><label class=\\"k-radio k-radio--base k-radio--base__dark k-radio--md k-cur-disabled\\"><input disabled=\\"\\" type=\\"radio\\" value=\\"false\\" hidden=\\"\\"> <div class=\\"k-radio--box k-radio--box__disabled animate-ikun-checking\\"></div> 1</label> <label class=\\"k-radio k-radio--base k-radio--base__dark k-radio--md k-cur-disabled\\"><input disabled=\\"\\" type=\\"radio\\" value=\\"false\\" hidden=\\"\\"> <div class=\\"k-radio--box k-radio--box__disabled animate-ikun-checking\\"></div> 2</label> <label class=\\"k-radio k-radio--base k-radio--base__dark k-radio--md k-cur-disabled\\"><input disabled=\\"\\" type=\\"radio\\" value=\\"true\\" hidden=\\"\\"> <div class=\\"k-radio--box k-radio--box__disabled animate-ikun-checking\\"><span class=\\"k-icon--base k-icon-transition i-carbon:radio-button-checked !text-white\\" role=\\"\\" aria-hidden=\\"true\\"></span></div> 3</label></div></div>"`;

exports[`Test: KRadioGroup > When it is disabled, the value of the radio can be initialized. 1`] = `"<div class=\\"h-full\\" style=\\"width: 600px;\\"><div class=\\"k-radio-group k-radio-group--base\\"><label class=\\"k-radio k-radio--base k-radio--base__dark k-radio--md k-cur-disabled\\"><input disabled=\\"\\" type=\\"radio\\" value=\\"true\\" hidden=\\"\\"> <div class=\\"k-radio--box k-radio--box__disabled animate-ikun-checking\\"><span class=\\"k-icon--base k-icon-transition i-carbon:radio-button-checked !text-white\\" role=\\"\\" aria-hidden=\\"true\\"></span></div> 1</label> <label class=\\"k-radio k-radio--base k-radio--base__dark k-radio--md k-cur-disabled\\"><input disabled=\\"\\" type=\\"radio\\" value=\\"false\\" hidden=\\"\\"> <div class=\\"k-radio--box k-radio--box__disabled animate-ikun-checking\\"></div> 2</label> <label class=\\"k-radio k-radio--base k-radio--base__dark k-radio--md k-cur-disabled\\"><input disabled=\\"\\" type=\\"radio\\" value=\\"false\\" hidden=\\"\\"> <div class=\\"k-radio--box k-radio--box__disabled animate-ikun-checking\\"></div> 3</label></div></div>"`;

exports[`Test: KRadioGroup > When it is disabled, when a value of the radio changes, the group value is not triggered. 1`] = `"<div class=\\"h-full\\" style=\\"width: 600px;\\"><div id=\\"group_trigger_value\\">3</div> <div class=\\"k-radio-group k-radio-group--base\\"><label class=\\"k-radio k-radio--base k-radio--base__dark k-radio--md k-cur-disabled\\"><input disabled=\\"\\" type=\\"radio\\" value=\\"false\\" hidden=\\"\\"> <div class=\\"k-radio--box k-radio--box__disabled animate-ikun-checking\\"></div> 1</label> <label class=\\"k-radio k-radio--base k-radio--base__dark k-radio--md k-cur-disabled\\"><input disabled=\\"\\" type=\\"radio\\" value=\\"false\\" hidden=\\"\\"> <div class=\\"k-radio--box k-radio--box__disabled animate-ikun-checking\\"></div> 2</label> <label class=\\"k-radio k-radio--base k-radio--base__dark k-radio--md k-cur-disabled\\"><input disabled=\\"\\" type=\\"radio\\" value=\\"true\\" hidden=\\"\\"> <div class=\\"k-radio--box k-radio--box__disabled animate-ikun-checking\\"><span class=\\"k-icon--base k-icon-transition i-carbon:radio-button-checked !text-white\\" role=\\"\\" aria-hidden=\\"true\\"></span></div> 3</label></div></div>"`;

exports[`Test: KRadioGroup > When it is not disabled, the value changes and the status of the radio changes. 1`] = `"<div class=\\"h-full\\" style=\\"width: 600px;\\"><button id=\\"group_change_btn\\"></button> <div class=\\"k-radio-group k-radio-group--base\\"><label class=\\"k-radio k-radio--base k-radio--base__dark k-radio--md\\"><input type=\\"radio\\" value=\\"true\\" hidden=\\"\\"> <div class=\\"k-radio--box k-radio__selected animate-ikun-checking\\"><span class=\\"k-icon--base k-icon-transition i-carbon:radio-button-checked !text-white\\" role=\\"\\" aria-hidden=\\"true\\"></span></div> 1</label> <label class=\\"k-radio k-radio--base k-radio--base__dark k-radio--md\\"><input type=\\"radio\\" value=\\"false\\" hidden=\\"\\"> <div class=\\"k-radio--box animate-ikun-checking\\"></div> 2</label> <label class=\\"k-radio k-radio--base k-radio--base__dark k-radio--md\\"><input type=\\"radio\\" value=\\"false\\" hidden=\\"\\"> <div class=\\"k-radio--box animate-ikun-checking\\"></div> 3</label></div></div>"`;

exports[`Test: KRadioGroup > When it is not disabled, the value of the radio can be initialized. 1`] = `"<div class=\\"h-full\\" style=\\"width: 600px;\\"><div class=\\"k-radio-group k-radio-group--base\\"><label class=\\"k-radio k-radio--base k-radio--base__dark k-radio--md\\"><input type=\\"radio\\" value=\\"false\\" hidden=\\"\\"> <div class=\\"k-radio--box animate-ikun-checking\\"></div> 1</label> <label class=\\"k-radio k-radio--base k-radio--base__dark k-radio--md\\"><input type=\\"radio\\" value=\\"true\\" hidden=\\"\\"> <div class=\\"k-radio--box k-radio__selected animate-ikun-checking\\"><span class=\\"k-icon--base k-icon-transition i-carbon:radio-button-checked !text-white\\" role=\\"\\" aria-hidden=\\"true\\"></span></div> 2</label> <label class=\\"k-radio k-radio--base k-radio--base__dark k-radio--md\\"><input type=\\"radio\\" value=\\"false\\" hidden=\\"\\"> <div class=\\"k-radio--box animate-ikun-checking\\"></div> 3</label></div></div>"`;

exports[`Test: KRadioGroup > When it is not disabled, when a value of the radio changes, the group value is triggered. 1`] = `"<div class=\\"h-full\\" style=\\"width: 600px;\\"><div id=\\"group_trigger_value\\">2</div> <div class=\\"k-radio-group k-radio-group--base\\"><label class=\\"k-radio k-radio--base k-radio--base__dark k-radio--md\\"><input type=\\"radio\\" value=\\"false\\" hidden=\\"\\"> <div class=\\"k-radio--box animate-ikun-checking\\"></div> 1</label> <label class=\\"k-radio k-radio--base k-radio--base__dark k-radio--md\\"><input type=\\"radio\\" value=\\"true\\" hidden=\\"\\"> <div class=\\"k-radio--box k-radio__selected animate-ikun-checking\\"><span class=\\"k-icon--base k-icon-transition i-carbon:radio-button-checked !text-white\\" role=\\"\\" aria-hidden=\\"true\\"></span></div> 2</label> <label class=\\"k-radio k-radio--base k-radio--base__dark k-radio--md\\"><input type=\\"radio\\" value=\\"false\\" hidden=\\"\\"> <div class=\\"k-radio--box animate-ikun-checking\\"></div> 3</label></div></div>"`;

exports[`Test: KRadioGroup > When the disabled value changes, the radio value can be synchronized. 1`] = `"<div class=\\"h-full\\" style=\\"width: 600px;\\"><button id=\\"group_change_value_btn\\"></button> <button id=\\"group_change_disabled_btn\\"></button> <div class=\\"k-radio-group k-radio-group--base\\"><label class=\\"k-radio k-radio--base k-radio--base__dark k-radio--md\\"><input type=\\"radio\\" value=\\"true\\" hidden=\\"\\"> <div class=\\"k-radio--box k-radio__selected animate-ikun-checking\\"><span class=\\"k-icon--base k-icon-transition i-carbon:radio-button-checked !text-white\\" role=\\"\\" aria-hidden=\\"true\\"></span></div> 1</label> <label class=\\"k-radio k-radio--base k-radio--base__dark k-radio--md\\"><input type=\\"radio\\" value=\\"false\\" hidden=\\"\\"> <div class=\\"k-radio--box animate-ikun-checking\\"></div> 2</label> <label class=\\"k-radio k-radio--base k-radio--base__dark k-radio--md\\"><input type=\\"radio\\" value=\\"false\\" hidden=\\"\\"> <div class=\\"k-radio--box animate-ikun-checking\\"></div> 3</label></div></div>"`;

exports[`Test: KRadioGroup > props: cls 1`] = `"<div class=\\"k-radio-group k-radio-group--base k-radio-group--test\\"></div>"`;

exports[`Test: KRadioGroup > props: size 1`] = `"<div id=\\"group-sm\\" class=\\"h-full\\" style=\\"width: 600px;\\"><div class=\\"k-radio-group k-radio-group--base\\"><label class=\\"k-radio k-radio--base k-radio--base__dark k-radio--sm\\"><input type=\\"radio\\" value=\\"false\\" hidden=\\"\\"> <div class=\\"k-radio--box animate-ikun-checking\\"></div> 1</label> <label class=\\"k-radio k-radio--base k-radio--base__dark k-radio--sm\\"><input type=\\"radio\\" value=\\"false\\" hidden=\\"\\"> <div class=\\"k-radio--box animate-ikun-checking\\"></div> 2</label> <label class=\\"k-radio k-radio--base k-radio--base__dark k-radio--sm\\"><input type=\\"radio\\" value=\\"false\\" hidden=\\"\\"> <div class=\\"k-radio--box animate-ikun-checking\\"></div> 3</label></div></div> <div id=\\"group-md\\" class=\\"h-full\\" style=\\"width: 600px;\\"><div class=\\"k-radio-group k-radio-group--base\\"><label class=\\"k-radio k-radio--base k-radio--base__dark k-radio--md\\"><input type=\\"radio\\" value=\\"false\\" hidden=\\"\\"> <div class=\\"k-radio--box animate-ikun-checking\\"></div> 1</label> <label class=\\"k-radio k-radio--base k-radio--base__dark k-radio--md\\"><input type=\\"radio\\" value=\\"false\\" hidden=\\"\\"> <div class=\\"k-radio--box animate-ikun-checking\\"></div> 2</label> <label class=\\"k-radio k-radio--base k-radio--base__dark k-radio--md\\"><input type=\\"radio\\" value=\\"false\\" hidden=\\"\\"> <div class=\\"k-radio--box animate-ikun-checking\\"></div> 3</label> <label class=\\"k-radio k-radio--base k-radio--base__dark k-radio--md\\"><input type=\\"radio\\" value=\\"false\\" hidden=\\"\\"> <div class=\\"k-radio--box animate-ikun-checking\\"></div> 4</label> <label class=\\"k-radio k-radio--base k-radio--base__dark k-radio--md\\"><input type=\\"radio\\" value=\\"false\\" hidden=\\"\\"> <div class=\\"k-radio--box animate-ikun-checking\\"></div> 5</label></div></div> <div id=\\"group-lg\\" class=\\"h-full\\" style=\\"width: 600px;\\"><div class=\\"k-radio-group k-radio-group--base\\"><label class=\\"k-radio k-radio--base k-radio--base__dark k-radio--lg\\"><input type=\\"radio\\" value=\\"false\\" hidden=\\"\\"> <div class=\\"k-radio--box animate-ikun-checking\\"></div> 1</label> <label class=\\"k-radio k-radio--base k-radio--base__dark k-radio--lg\\"><input type=\\"radio\\" value=\\"false\\" hidden=\\"\\"> <div class=\\"k-radio--box animate-ikun-checking\\"></div> 2</label></div></div>"`;
